activerecord/lib/active_record/insert_rows.rb

```diff
@@ -0,0 +1,145 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/enumerable"
+
+module ActiveRecord
+  class InsertRows # :nodoc:
+    attr_reader :model, :connection, :inserts, :columns, :returning
+
+    class << self
+      def execute(model, ...)
+        model.with_connection do |c|
+          new(model, c, ...).execute
+        end
+      end
+    end
+
+    def initialize(model, connection, inserts, columns:, returning: nil)
+      @model = model
+      @connection = connection
+
+      duplicated_columns = columns.group_by(&:itself).select { |_, v| v.size > 1 }.keys
```

Comment:
```suggestion
      duplicated_columns = columns.tally.select { |_, c| c > 1 }.keys
```

Reply:
We might also want to cast them to string, so that `[:foo, "foo"]` is properly seen as a duplicate.